### ISSUE TITLE
Support custom shard keys

### ISSUE THREAD

**Issue by coffeemug** [rethinkdb/rethinkdb#1862]
Since we'll be revamping the clustering infrastructure, it ought to be possible to partition documents into different datacenters based on some metric.

**Comment by no-more** [comment]
Hello,

I was looking exactly for this feature, I absolutely need to enforce the data localization per region (America, Europe...). This is a top requirement for my project, I've tried to ask for this here (https://groups.google.com/forum/#!topic/rethinkdb/HknP-wkX-Nc) before seeing this issue.

Of course this might not be very easy to implement, but I really thing it's a top feature for more and more international developments.

I really hope this would get implemented soon.

**Comment by danielmewes** [comment]
@no-more for your particular use case, the somewhat simpler https://github.com/rethinkdb/rethinkdb/issues/5431 would be sufficient I think. You would just need to make sure that your application code generates primary keys with a region prefix (e.g. `r.expr("America-").add(r.uuid())`).

**Comment by thelinuxlich** [comment]
Something for inspiration: http://redis.io/topics/cluster-spec#keys-hash-tags

**Comment by no-more** [comment]
Hi,
thanks but I don't understand how/where I can define that a key starting with america have to be in serveur tag america ?

If I have a big cluster does `r.expr("America-").add(r.uuid()))` ensure uniquity ?

Thanks.

**Comment by danielmewes** [comment]
@no-more Indeed, I meant if #5431 was implemented, then you could configure one (or multiple) shards explicitly to contain all the "America-*" keys. Currently this is not directly possible.

`r.uuid()` has a negligibly small chance of collision, and for practical purposes is sufficient to ensure uniquity. Furthermore RethinkDB actually checks that primary keys are unique, so in the (extremely unlikely) case that a collision is generated, the `insert` query will simply fail with a duplicate key error and you can retry it.
